plugins/modules/mso_schema_site_contract_service_graph_listener.py

```diff
@@ -0,0 +1,744 @@
+    listener_protocol = module.params.get("listener_protocol")
+    listener_port = module.params.get("listener_port")
+    security_policy = LISTENER_SECURITY_POLICY_MAP.get(module.params.get("security_policy"))
+    ssl_certificates = module.params.get("ssl_certificates")
+    frontend_ip = module.params.get("frontend_ip")
+    rules = module.params.get("rules")
+    state = module.params.get("state")
+
+    if listener_protocol == "https":
+        mso.input_validation("listener_protocol", "https", ["security_policy", "ssl_certificates"], module.params, None, object_name=listener)
+
+    mso_schema = MSOSchema(mso, schema, template)
+    mso_schema.set_template(template)
+    schema_id = mso.lookup_schema(schema, True)
+    mso_schema.set_site(template, site)
+    mso_schema.set_site_contract(contract, False)
+
+    service_graph_ref = mso_schema.schema_objects["site_contract"].details.get("serviceGraphRelationship", {}).get("serviceGraphRef")
+    service_nodes = mso_schema.schema_objects["site_contract"].details.get("serviceGraphRelationship", {}).get("serviceNodesRelationship", [])
+
+    listeners_data = []
+    parent_present = False
+
+    if service_graph_ref is None:
+        mso.fail_json(msg="The site contract: {0} is not associated with a service graph.".format(contract))
+    # Parent object present check
+    else:
+        if service_node_index is not None and service_node_index >= 0:
+            if len(service_nodes) > 0 and len(service_nodes) > service_node_index:
+                service_node = service_nodes[service_node_index]
+                # Query all listeners under a contract service node
+                # The below condition was never false if the service graph was created properly, but condition is required to avoid error
+                if service_node.get("serviceNodeRef") and service_node.get("serviceNodeRef").split("/")[-1] == "node{0}".format(service_node_index + 1):
+                    listeners = service_node.get("deviceConfiguration", {}).get("cloudLoadBalancer", {}).get("listeners", [])
+                    if listeners:
+                        parent_present = True
+                        for listener_data in listeners:
+                            if listener_data.get("name") == listener:
+                                listeners_data.append(listener_data)
+                                break
+                else:
+                    mso.fail_json(
+                        msg="The service_node_index: {0} is not matching with the service node reference: {1}.".format(
+                            service_node_index, service_node.get("serviceNodeRef")
+                        )
+                    )
+            else:
+                mso.fail_json(msg="The service_node_index: {0} is out of range.".format(service_node_index))
+
+        # Query all listeners under a contract does not require service_node_index, so the below condition is required
+        elif state == "query":
```

Comment:
Cant we separate this from the lookup logic? So the code structure would be something like:

1. init module and param
2. search object in existing config -> exists store in variable and mso.existing else None
3. query/present/absent logic which can use the found object